### ISSUE TITLE
Decide on FMS compile flags

### ISSUE THREAD

**Issue by harshula** [ACCESS-NRI/FMS#1]
The existing CMakeLists.txt contains the following compile flags: https://github.com/ACCESS-NRI/FMS/blob/38c7dbef9424b8b263820c7e189a4458f049db78/CMakeLists.txt#L40-L64

The upstream FMS compile flags are: https://github.com/ACCESS-NRI/FMS/commit/02bdd2fb831d197a4a62861ba7ef762bd05528a8

The MOM5 compile flags are: https://github.com/ACCESS-NRI/MOM5/blob/6d8768648ffa2326b4817c3de9ccb25e9d808585/bin/mkmf.template.nci#L39-L51

Which of them should we use?

**Comment by harshula** [comment]
Example of compilation flags from MOM5 with git subtree FMS (`spack install mom5@master %intel@2021.2.0 target=x86_64`): `mpifort -Duse_netCDF -Duse_libMPI -DACCESS_OM -DUSE_OCEAN_BGC  -fpp -Wp,-w -I/opt/release/linux-rocky8-x86_64/intel-2021.2.0/oasis3-mct-master-df2rx2hjfd2xmwfu34nzfkdugo34voew/include/psmile.MPI1 -I/opt/release/linux-rocky8-x86_64/intel-2021.2.0/oasis3-mct-master-df2rx2hjfd2xmwfu34nzfkdugo34voew/include -I/opt/release/linux-rocky8-x86_64/intel-2021.2.0/libaccessom2-master-ss24s7pyzzhbzck5dsqjd2z4ca3lavkw/include -I/opt/release/linux-rocky8-x86_64/intel-2021.2.0/netcdf-fortran-4.6.1-dutskkuweiulkdk3vmdzvmshgc4wtuo2/include -fno-alias -safe-cray-ptr -fpe0 -ftz -assume byterecl -i4 -r8 -nowarn -check noarg_temp_created -assume nobuffered_io -convert big_endian -grecord-gcc-switches -align all -fp-model precise -fp-model source -align all -g3 -O2 -xCORE-AVX2 -debug all -check none -traceback  -c -I/tmp/root/spack-stage/spack-stage-mom5-master-vgak75luzyoz6ietfh46jpj2laox6xxf/spack-src/src/shared/include	/tmp/root/spack-stage/spack-stage-mom5-master-vgak75luzyoz6ietfh46jpj2laox6xxf/spack-src/src/shared/fms/fms.F90
`

**Comment by harshula** [comment]
Looks like `src/shared/fms/fms.F90` is currently compiled with MOM5 compile flags. So, if we want to be consistent with existing MOM5 compilations, I guess we have to build `FMS` with MOM5 compile flags?

**Comment by dougiesquire** [comment]
> I guess we have to build FMS with MOM5 compile flags?

Agreed - I think we want to consistent with what was being done before

**Comment by harshula** [comment]
Hi @manodeep ,

What compiler flags should we use for `IntelLLVM`? We could start with with FMS upstream's compiler flags:

cmake/compiler_flags_IntelLLVM_C.cmake:
```
set( CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -traceback")
set( CMAKE_C_FLAGS_RELEASE "-qno-opt-dynamic-align -O2 -debug minimal")
set( CMAKE_C_FLAGS_DEBUG "-O0 -g -ftrapuv")
set( CMAKE_C_LINK_FLAGS "")
```
cmake/compiler_flags_IntelLLVM_Fortran.cmake:
```
set(r4_flags "-real-size 32") # Fortran flags for 32BIT precision
set(r8_flags "-real-size 64") # Fortran flags for 64BIT precision
set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -fpp -fno-alias -auto -safe-cray-ptr -ftz -assume byterecl -align array64byte -nowarn -traceback")
set(CMAKE_Fortran_FLAGS_RELEASE "-O2 -debug minimal -nowarn -qoverride-limits -qno-opt-dynamic-align")
set(CMAKE_Fortran_FLAGS_DEBUG "-g -O0 -check -check noarg_temp_created -check nopointer -warn -warn noerrors -fpe0 -ftrapuv")
set(CMAKE_Fortran_LINK_FLAGS "")
```

**Comment by manodeep** [comment]
Those flags look fine to me. Was there a reason why the oneAPI compiler returns a fatal error?

**Comment by harshula** [comment]
Hi @manodeep 

Have a look at https://github.com/ACCESS-NRI/FMS/blob/development/CMakeLists.txt and see what happens when `CMAKE_Fortran_COMPILER_ID` is `IntelLLVM`.

I should be able to get back to FMS this week and add `IntelLLVM` support.

Can you please help me by checking each of the compiler flags in https://github.com/ACCESS-NRI/FMS/issues/1#issuecomment-2306875467 and tell me if any are incompatible with IFX? If they are, please suggest an alternative.

**Comment by manodeep** [comment]
> Hi [@manodeep](https://github.com/manodeep)
> 
> Have a look at https://github.com/ACCESS-NRI/FMS/blob/development/CMakeLists.txt and see what happens when `CMAKE_Fortran_COMPILER_ID` is `IntelLLVM`.
> 

@harshula Uhh `cmake` returns a fatal error [here](https://github.com/ACCESS-NRI/FMS/blob/development/CMakeLists.txt#L83) - which is what I was saying[here](https://github.com/ACCESS-NRI/FMS/issues/1#issuecomment-2673443538). It would have made sense to treat the oneAPI as the same as classic - do you know why the choice was made to return a "fatal error"?

> I should be able to get back to FMS this week and add `IntelLLVM` support.
> 
Fantastic! If it would be helpful, I am happy to jump on a zoom call and try to compile those oneAPI packages out together. 

> Can you please help me by checking each of the compiler flags in [#1 (comment)](https://github.com/ACCESS-NRI/FMS/issues/1#issuecomment-2306875467) and tell me if any are incompatible with IFX? If they are, please suggest an alternative.

Those flags are compatible with `ifx`. `-ip` is a classic Intel flag that I know is not compatible with `ifx`. Another flag that I suspect is no longer supported is `-mp1`  - which is used in UM7; possibly elsewhere.

I would *really* like to add `-flto` to the compiler flags - the reports are that the Intel optimisations are *only* enabled when `-flto` is set; otherwise, you get the default `LLVM` optimiser. See this [comment](https://community.intel.com/t5/Intel-Fortran-Compiler/ifx-compiler-options/m-p/1584388/highlight/true#M171448)

**Comment by manodeep** [comment]
Confirming that `-mp1` is indeed not supported by `ifx` - got this warning while compiling `UM7` with oneAPI `ifx: command line warning #10148: option '-mp1' not supported`

**Comment by harshula** [comment]
Hi @manodeep , git tag `dev_2025.02.1` contains CMakelists.txt with upstream compiler flags and it builds successfully. Using the MOM5 intel Fortran compiler flags for oneapi is failing to build because `MPI_Fortran` can not be found.

Can you please document all your findings, including what you have documented in this issue, in https://github.com/ACCESS-NRI/dev-docs/wiki/ ?
